mxclient/public-rooms.go

```diff
@@ -71,12 +72,29 @@ func (r *WorldReadableRooms) Update() error {
 	return nil
 }
 
-// For future when we support filtering the public room directory (LOCALLY)
-//func (r *WorldReadableRooms) GetFilteredPage(page, pageSize int, query string) []gomatrix.PublicRoom {
-//	r.roomsMutex.RLock()
-//	defer r.roomsMutex.RUnlock()
-//	return nil
-//}
+// GetFilteredPage returns a filtered & paginated slice of the WorldReadableRooms Collection
+func (r *WorldReadableRooms) GetFilteredPage(page, pageSize int, query string) []gomatrix.PublicRoom {
+	r.roomsMutex.RLock()
+	defer r.roomsMutex.RUnlock()
+
+	lowerQuery := strings.ToLower(query)
+
+	filteredRooms := make([]gomatrix.PublicRoom, 0, pageSize)
+	for _, room := range r.rooms {
+		if len(filteredRooms) > pageSize {
```

Comment:
My feeling is this maybe wants to be `>=`, otherwise `filteredRooms` can overflow `pageSize` by 1?